### PR TITLE
codec_adapter: cadence: fix reset procedure + improve prepare

### PR DESCRIPTION
**Pull request by mrajwa**
This patch set:
- improves prepare procedure by adding INIT_PROCESS and INIT_DONE API calls
- fixes the reset procedure to address https://github.com/thesofproject/sof/issues/4104
   Cadence documentation doesn't document reset procedure at all so for a long time we were just freeing all the memory codec allocated and start over. However this is very inefficient way of performing simple reset of parameters. Hence this patch  https://github.com/thesofproject/sof/pull/4162/files tried to simplify it and while it does reset parameters it also breaks the flow since XA_CMD_TYPE_INIT_API_PRE_CONFIG_PARAMS is only allowed during INIT phase. After code inspection of the codec library XA_CMD_TYPE_INIT_PROCESS seems to do what we need to perform the reset and hence this patch relies on that API call.
- minor cleanup of wrong comments

### CODE REVIEW

File: src/audio/codec_adapter/codec/generic.c
Comment (lgirdwood) [inline src/audio/codec_adapter/codec/generic.c:342]: Not following, if we reset() the codec then any configuration via IPC will also be reset like channels, rate, format etc ?

Comment (mrajwa) [reply src/audio/codec_adapter/codec/generic.c:342]: Yes it will be reset to their default values.

File: src/audio/codec_adapter/codec/cadence.c
Comment (dbaluta) [inline src/audio/codec_adapter/codec/cadence.c:455]: I don't think this will work. 

``cadence_codec_init_prcess`` will call ``API_CALL(cd, XA_API_CMD_INIT, XA_CMD_TYPE_INIT_PROCESS, NULL, ret);`` 

This will fail because at cadence_codec_prepare() time there is no data in the input buffer. Does this work for Intel DAP codec?

Comment (mrajwa) [reply src/audio/codec_adapter/codec/cadence.c:455]: @dbaluta but do you think `XA_CMD_TYPE_INIT_PROCESS` requires data in the input buffer?

Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:455]: @mrajwa yes. Here is what DAP VLLDP Programmers Guide says:

image

Same for mp3, aac or any other cadence codec. That's why i added the cadence_codec_init_process() call inside the copy() function

Comment (mrajwa) [reply src/audio/codec_adapter/codec/cadence.c:455]: We need to clarify this with cadence, namely what is the correct way of performing the reset - this is nowhere defined. I noticed INIT_PROCESS does reset the parameters by DAP lib code inspection.

> Does this work for Intel DAP codec?

Yes it works for DAP

Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:455]: Ok, i see. It doesn't work for mp3 for example. It needs to have some input data in the buffers. Care to start a thread with Cadence support?
I will be on vacation starting tomorrow returning on June, 9 but will still check my emails.

Comment (mrajwa) [reply src/audio/codec_adapter/codec/cadence.c:455]: Sure, no problem

Comment (lgirdwood) [reply src/audio/codec_adapter/codec/cadence.c:455]: @dbaluta @mrajwa what's the plan for merging ? do we want to merge the other patches in this PR and wait until clarification from Cadence on this one ?

Comment (keyonjie) [reply src/audio/codec_adapter/codec/cadence.c:455]: Hi @mrajwa @dbaluta as fixing of the #4104 is still in our v1.8 milestone which should be code freeze by next week, please let us know about any update and if the fix can get aligned and merged before that.

Comment (mrajwa) [reply src/audio/codec_adapter/codec/cadence.c:460]: @lgirdwood @keyonjie patch updated. I would like to ask @kkarask to verify if it fixes the issue https://github.com/thesofproject/sof/issues/4104

Comment (lgirdwood) [reply src/audio/codec_adapter/codec/cadence.c:460]: @paulstelian97 @dbaluta @iuliana-prodan can someone check with MP3

**Review by lgirdwood — APPROVED** [review]

File: src/audio/codec_adapter/codec/cadence.c
Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:460]: @lgirdwood will verify them today.

Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:460]: @mrajwa @lgirdwood the solution is pretty clever. Anyhow, for some strange reasons:

```
+       API_CALL(cd, XA_API_CMD_INIT, XA_CMD_TYPE_INIT_DONE_QUERY,
+                &codec->cpd.init_done, ret);
```

makes codec->cpd.init = 1 and ret = 0!! And this will cause MP3 initialization to be skipped in copy function. 

I think either we have in the API or do we need to do additional steps before acquiring the status. I will send an email to Cadence to ask for this. 

Meanwhile, I think we can also make a per codec function that will initialize codec->cpd.init = 0 or 1.

@lgirdwood @mrajwa will add you in the discussion with Cadence.

Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:460]: > Hmm and the playback fails at process() then? If so this is weird and suggest me a codec bug, let's face it, codec returns INIT_DONE but then fails due to uninitialized codec?

Indeed this happens!

Comment (mrajwa) [reply src/audio/codec_adapter/codec/cadence.c:460]: Can you, for the testing purposes call the codec API with `XA_CMD_TYPE_DO_RUNTIME_INIT` command before the update of init_done at the end of the prepare() function? I have a feeling this may update the flag for compress streams.

Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:460]: @mrajwa Ack. Will get back to you when get some time for testing.

Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:460]: @mrajwa I checked with your suggestion. But no luck. The init_done is still set at 1.

```
      API_CALL(cd, XA_API_CMD_EXECUTE, XA_CMD_TYPE_DO_RUNTIME_INIT, NULL, ret);
     API_CALL(cd, XA_API_CMD_INIT, XA_CMD_TYPE_INIT_DONE_QUERY, &codec->cpd.init_done, ret);
```
I checked with Cadence support team and they confirm that this happens and might be a library bug. They will forward the problem to engineering. 

Anyhow, what is exactly our problem to fix with this reset procedure. I'm sorry I'm still after vacation and is hard for me to navigate through the sea of emails. Can you provide a simple description of the problem we are trying to solve. 

On another note, I'm having a problem with second cplay run. It's always get stuck. Anyhow, if I leave enough time between 1st run and 2nd run of cplay everything works (system runtime suspend/resume get into action).

Cc: @paulstelian97

Comment (lgirdwood) [reply src/audio/codec_adapter/codec/cadence.c:460]: > On another note, I'm having a problem with second cplay run. It's always get stuck. Anyhow, if I leave enough time between 1st run and 2nd run of cplay everything works (system runtime suspend/resume get into action).

Sounds like kernel runtime PM is rebooting FW and reseting the state ?

**Comment by lgirdwood** [comment]
@mrajwa @dbaluta I will merge unless any objections ? We can figure out when Cadence get back to us.

**Comment by mrajwa** [comment]
@lgirdwood yes I am ok with that. The bug only affects compress streams.

File: src/audio/codec_adapter/codec/cadence.c
Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:460]: @lgirdwood true.

**Comment by dbaluta** [comment]
@lgirdwood please don't merge this. This breaks existing working usecases.

Actually, only the last commit ( codec_adapter: cadence: get init_done status on prepare ) is problematic. This breaks compress streams. 
I am also looking at this to find a better fix.
@mrajwa can you explain in a simple usecase what is the problem with existing code? I will try to replicate the bug on compress streams and fix it.

**Review by dbaluta — CHANGES_REQUESTED** [review]
@lgirdwood @mrajwa leave us time until Wednesday to fix it.

**Comment by mrajwa** [comment]
> @mrajwa can you please confirm that the problem we are trying to fix is that running aplay/cplay multiple time fails at the second run?

Yes that's about it, plus keep in mind real time parameters like i.e post gain shall be reset to their default values as well. Current implementation allows for consecutive play but  it doesn't reset parameters as it should.
We struggle with reset for quite a bit.. the solution proposed in this PR is trying to cover missing API in cadence codec but it turns our there is a bug in their logic as well (affects compress steams only). 

> I am also looking at this to find a better fix.

Well, it's not use who should come up with a fix but Cadence... What did they tell you, are they working on to fix that?

**Comment by dbaluta** [comment]
@mrajwa they forwarded the problem to engineering and they are working for a fix. Will send them another mail to ask for status.

**Comment by mrajwa** [comment]
@dbaluta okay, I would propose them dedicated API for parameters reset - the fact that we need to free all codec memory and allocate again shows the API is defective in the first place.

**Comment by dbaluta** [comment]
@mrajwa I agree. Can you continue the discussion with Cadence on email via thread stared by me?

**Comment by mrajwa** [comment]
@dbaluta sure but I didn't receive any thread, could you please forward?

**Comment by lgirdwood** [comment]
Ok, lets delay this until Cadence provide us with guidance. We can make this a hotfix when we are all aligned with each other and Cadence.

**Comment by dbaluta** [comment]
> @dbaluta sure but I didn't receive any thread, could you please forward?

@mrajwa I have added you to the thread since the first email with your linux.intel.com address. Now I also forwarded you the emails using your intel.com address. Let me know if you got them.

**Comment by mrajwa** [comment]
@dbaluta I have just updated with temporary solution for the init done bit. Is that OK for you?

File: src/audio/codec_adapter/codec/cadence.c
Comment (lyakh) [inline src/audio/codec_adapter/codec/cadence.c:460]: side note unrelated to this PR: I think it would be good to switch `API_CALL()` to
```
#define API_CALL(cd, cmd, sub_cmd, value) \
		(cd)->api((cd)->self, cmd, sub_cmd, value)
```
and then do explicitly
```
ret = API_CALL(cd, cmd, sub_cmd, value);
```

**Comment by dbaluta** [comment]
> @dbaluta I have just updated with temporary solution for the init done bit. Is that OK for you?

@mrajwa all good from my side now! Ready to merge.

**Review by dbaluta — APPROVED** [review]

File: src/audio/codec_adapter/codec/cadence.c
Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:460]: @lyakh not a bad idea, but I don't see a draw back from the current implementation either. It is just a mater of taste. We all know that API_CALL is a macro. 

Remember `spin_lock_irqsave` in the Linux kernel?

**Pull request merged**
